android/src/main/kotlin/com/builttoroam/devicecalendar/CalendarDelegate.kt

```diff
@@ -281,40 +283,40 @@ class CalendarDelegate : PluginRegistry.RequestPermissionsResultListener {
             val eventsUri = eventsUriBuilder.build()
             val eventsCalendarQuery = "(${Events.CALENDAR_ID} = $calendarId)"
             val eventsNotDeletedQuery = "(${Events.DELETED} != 1)"
-            val eventsIdsQueryElements = eventIds.map { "(${CalendarContract.Instances.EVENT_ID} = $it)" }
-            val eventsIdsQuery = eventsIdsQueryElements.joinToString(" OR ")
+            val eventsIdsQuery ="(${CalendarContract.Instances.EVENT_ID} IN (${eventIds.joinToString()}))"
 
             var eventsSelectionQuery = "$eventsCalendarQuery AND $eventsNotDeletedQuery"
-            if (eventsIdsQuery.isNotEmpty()) {
+            if (eventIds.isNotEmpty()) {
                 eventsSelectionQuery += " AND ($eventsIdsQuery)"
             }
-            val eventsSortOrder = Events.DTSTART + " ASC"
+            val eventsSortOrder = Events.DTSTART + " DESC"
+
             val eventsCursor = contentResolver?.query(eventsUri, EVENT_PROJECTION, eventsSelectionQuery, null, eventsSortOrder)
 
             val events: MutableList<Event> = mutableListOf()
 
             try {
-                if (eventsCursor?.moveToFirst() == true) {
-                    do {
+                GlobalScope.launch(Dispatchers.IO) {
+                    while (eventsCursor?.moveToNext() == true) {
                         val event = parseEvent(calendarId, eventsCursor) ?: continue
                         events.add(event)
-
-                    } while (eventsCursor.moveToNext())
+                    }
 
                     for (event in events) {
                         val attendees = retrieveAttendees(event.eventId!!, contentResolver)
                         event.organizer = attendees.firstOrNull { it.isOrganizer != null && it.isOrganizer }
                         event.attendees = attendees
                         event.reminders = retrieveReminders(event.eventId!!, contentResolver)
                     }
+                }.invokeOnCompletion {
+                    eventsCursor?.close()
+                    _registrar!!.activity().runOnUiThread {
+                        finishWithSuccess(_gson?.toJson(events), pendingChannelResult)
+                    }
                 }
             } catch (e: Exception) {
                 finishWithError(GENERIC_ERROR, e.message, pendingChannelResult)
-            } finally {
-                eventsCursor?.close()
```

Comment:
If exceptions happen within the globalscope, does the exception get propagated out to the wrapper try-catch? and are we able to force the eventsCursor to be closed?

Reply:
I've raised a new issue to track adding better error handling